pkg/capture/capture_manager.go

```diff
@@ -405,43 +410,42 @@ func (cm *Manager) rotate(ctx context.Context, writeoutChan chan<- capturetypes.
 		return
 	}
 
-	var rg RunGroup
+	// Iteratively rotate all interfaces. Since the rotation results are put on the writeoutChan for
+	// writeout by the DBWriter (which is sequential and certainly slower than the actual in-memory rotation)
+	// there is no significant benefit from running the rotations in parallel, thus allowing us to minimize
+	// congestion _and_ use a single shared local memory buffer
 	for _, iface := range ifaces {
-		mc, exists := cm.captures.Get(iface)
-		if exists {
-			rg.Run(func() {
+		if mc, exists := cm.captures.Get(iface); exists {
 
-				runCtx := withIfaceContext(ctx, mc.iface)
-				logger := logging.FromContext(runCtx)
+			runCtx := withIfaceContext(ctx, mc.iface)
+			logger, lockStart := logging.FromContext(runCtx), time.Now()
 
-				// Lock the running capture and perform the rotation
-				mc.lock()
+			// Lock the running capture in order to safely perform rotation tasks
+			mc.lock()
 
-				rotateResult := mc.rotate(runCtx)
+			// Extract capture stats in a separate goroutine to minimize rotation duration
+			statsRes := mc.fetchStatusInBackground(runCtx)
 
-				// log errors
-				// TODO: remove, once we expose error information more effectively
-				if len(mc.errMap) > 0 {
-					logger.With("error_map", mc.errMap).Debug("rotation errors")
-				}
+			// Perform the rotation
+			rotateResult := mc.rotate(runCtx)
 
-				// Since the capture is locked we can safely extract the (capture) status
-				// from the individual interfaces (and unlock no matter what)
-				stats, err := mc.status()
-				if err != nil {
-					logger.Errorf("failed to get capture stats: %v", err)
-				}
-				mc.unlock()
+			// log errors
+			// TODO: remove, once we expose error information more effectively
+			if len(mc.errMap) > 0 {
+				logger.With("error_map", mc.errMap).Debug("rotation errors")
+			}
 
-				writeoutChan <- capturetypes.TaggedAggFlowMap{
-					Map:   rotateResult,
-					Stats: *stats,
-					Iface: mc.iface,
-				}
-			})
+			stats := <-statsRes
+			mc.unlock()
+			logger.With("elapsed", time.Since(lockStart).Round(time.Microsecond).String()).Debug("interface locked")
```

Comment:
@els0r Here I'd like your opinion: I'd really like to have a way to get an idea of the time an interface is locked (not the global rotation time). From my point of view this is easiest with such a debug log statement, but metrics would be more pretty of course. However, I recall you saying that cardinality might be a problem, hence I'm not sure if that's a viable option (because of course we need to know which interface is affected). What's your take on this (keeping in mind that it might also be feasible to drop this entirely in case we don't even see any drops anymore :stuck_out_tongue: ).

Reply:
While metrics would be optimal in order to continuously measure this per interface, I would also go with a simple log line here. It's after all something you want to investigate ad-hoc rather than baseline for tool health.

Reply:
Sounds good to me, thx for the feedback!